libs/beacon/src/beacon_setup_service.cpp

```diff
@@ -128,238 +169,771 @@ BeaconSetupService::State BeaconSetupService::OnIdle()
 
-  // Creating member contributions
-  beacon_->manager.GenerateContribution();
-
-  // Pre-paring to send
-  for (auto &counter_party : beacon_->aeon.members)
+  if (callback_function_)
   {
-    DeliveryDetails details;
-    details.was_delivered                  = false;
-    share_delivery_details_[counter_party] = details;
+    callback_function_(beacon_);
   }
 
-  return State::SEND_SHARES;
+  beacon_.reset();
+  pre_dkg_rbc_.ResetCabinet({});
+  rbc_.ResetCabinet({});
+  connections_.clear();
+  ready_connections_.clear();
+  complaints_manager_.ResetCabinet(0);
+  complaints_answer_manager_.ResetCabinet(0);
+  qual_complaints_manager_.Reset();
+  shares_received_.clear();
+  coefficients_received_.clear();
+  qual_coefficients_received_.clear();
+  reconstruction_shares_received_.clear();
+
+  return State::IDLE;
 }
 
-BeaconSetupService::State BeaconSetupService::SendShares()
+/**
+ * Sends DKG message via reliable broadcast channel in dkg_service
+ *
+ * @param env DKGEnvelope containing message to the broadcasted
+ */
+void BeaconSetupService::SendBroadcast(DKGEnvelope const &env)
 {
-  // Getting information common to all counter parties
-  {
-    std::lock_guard<std::mutex> lock(mutex_);
-    auto                        verification_vector = beacon_->manager.GetVerificationVector();
-    auto                        from                = beacon_->manager.identity();
+  DKGSerializer serialiser;
+  serialiser << env;
+  rbc_.Broadcast(serialiser.data());
+}
 
-    for (auto &delivery_info : share_delivery_details_)
+/**
+ * Randomly initialises coefficients of two polynomials, computes the coefficients and secret
+ * shares and sends to cabinet members
+ */
+void BeaconSetupService::BroadcastShares()
+{
+  beacon_->manager.GenerateCoefficients();
+  SendBroadcast(DKGEnvelope{CoefficientsMessage{static_cast<uint8_t>(State::WAIT_FOR_SHARE),
+                                                beacon_->manager.GetCoefficients(), "signature"}});
+  for (auto &cab_i : beacon_->aeon.members)
+  {
+    if (cab_i == identity_)
     {
-      auto &details = delivery_info.second;
-      if (details.was_delivered)
-      {
-        continue;
-      }
+      continue;
+    }
+    std::pair<MessageShare, MessageShare> shares{beacon_->manager.GetOwnShares(cab_i.identifier())};
 
-      auto counter_party = delivery_info.first;
-      auto share         = beacon_->manager.GetShare(counter_party);
+    fetch::serializers::SizeCounter counter;
+    counter << shares;
 
-      if (counter_party == identity_)
-      {
-        ShareSubmission submission;
-        submission.from                = from;
-        submission.share               = std::move(share);
-        submission.verification_vector = verification_vector;
+    fetch::serializers::MsgPackSerializer serializer;
+    serializer.Reserve(counter.size());
+    serializer << shares;
+    endpoint_.Send(cab_i.identifier(), SERVICE_DKG, CHANNEL_SECRET_KEY, serializer.data());
```

Comment:
This is the point we want to be encrypted right? If so I'll chat to Ed

Reply:
Yes, the secret shares need to be encrypted.